apis/common/parent/project.go

```diff
@@ -0,0 +1,132 @@
+package parent
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ Parent = &ProjectParent{}
+
+type ProjectParent struct {
+	ProjectID string
+}
+
+var _ ParentBuilder = &ProjectRef{}
+
+// Project specifies the resource's GCP hierarchy (Project/Folder/Organization).
+// +kubebuilder:object:generate:=true
+type ProjectRef struct {
+	/* The `projectID` field of a project, when not managed by Config Connector. */
+	External string `json:"external,omitempty"`
+	/* The `name` field of a `Project` resource. */
+	Name string `json:"name,omitempty"`
+	/* The `namespace` field of a `Project` resource. */
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// Builds a the ProjectParent from ProjectRef.
+// If `projectRef.external` is given, parse projectID from External, otherwise find the ConfigConnector project
+// according to `projectRef.name` and `projectRef.namespace`.
+func (p *ProjectRef) Build(ctx context.Context, reader client.Reader, othernamespace string, parent Parent) error {
+	projectParent, ok := parent.(*ProjectParent)
+	if !ok {
+		return fmt.Errorf("build invalid parent, except %T", &ProjectParent{})
+	}
+	if p.External != "" {
+		if p.Name != "" {
+			return fmt.Errorf("cannot specify both name and external on project reference")
+		}
+
+		tokens := strings.Split(p.External, "/")
+		if len(tokens) == 1 {
+			projectParent.ProjectID = tokens[0]
+			return nil
+		}
+		if len(tokens) == 2 && tokens[0] == "projects" {
+			projectParent.ProjectID = tokens[1]
+			return nil
+		}
+		return fmt.Errorf("format of project external=%q was not known (use projects/<projectId> or <projectId>)", p.External)
+	}
+
+	if p.Name == "" {
+		return fmt.Errorf("must specify either name or external on project reference")
+	}
+
+	key := types.NamespacedName{
+		Namespace: p.Namespace,
+		Name:      p.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = othernamespace
+	}
+
+	project := &unstructured.Unstructured{}
+	project.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "resourcemanager.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "Project",
+	})
+	if err := reader.Get(ctx, key, project); err != nil {
+		if apierrors.IsNotFound(err) {
+			return fmt.Errorf("referenced Project %v not found", key)
+		}
+		return fmt.Errorf("error reading referenced Project %v: %w", key, err)
+	}
+
+	projectID, _, err := unstructured.NestedString(project.Object, "spec", "resourceID")
+	if err != nil {
+		return fmt.Errorf("reading spec.resourceID from %v %v/%v: %w", project.GroupVersionKind().Kind, p.Namespace, p.Name, err)
+	}
+	if projectID == "" {
+		projectID = project.GetName()
+	}
+	projectParent.ProjectID = projectID
+	return nil
+}
+
+func ParseProjectParent(external string) (*ProjectParent, error) {
```

Comment:
We _could_ make this a method on FooIdentity, something like `ParseExternal() error`